### ISSUE TITLE
[Validator] Error on validating STU3 profile that extends a string/code/markdown type

### ISSUE THREAD

**Issue by mbaltus** [FirelyTeam/fhir-net-api#1140]
When you have an STU3 profile with an extension on one of the string types, validating that profile results in a warning and error on eld-13:

[WARNING] Evaluation of FhirPath for constraint 'eld-13' failed: Invocation of 'isDistinct' failed: Invocation of 'binary.&' failed: Invocation of 'binary.&' failed: Cannot cast from 'any single value' to 'String' (at StructureDefinition.snapshot[0].element[40])      
[ERROR] Instance failed constraint eld-13 "Types must be unique by the combination of code and profile"(further diagnostics: type.select(code&profile&targetProfile).isDistinct()) (at StructureDefinition.snapshot[0].element[40])

Steps to reproduce:
- Create a profile on Observation
- Extend the status field with for example the translation extension (http://hl7.org/fhir/StructureDefinition/translation)
- Save the profile with a snapshot component
- Validate the resulting SD to see the error

**Comment by mbaltus** [comment]
Sorry, this is actually a duplicate of #874

**Comment by wmrutten** [comment]
What is the status of this issue? Has this been fixed in the latest API release for STU3?
Maybe you can review the validation on Simplifier.

**Comment by ewoutkramer** [comment]
I have reproduced the issue - partially. I do not get the FhirPath execution errors anymore (which can be attributed to some of the improvements to the FhirPath engine since 1.3) - but indeed this eld-13 is still triggered.

Part of the problem is that the `element` the fhirpath fails on looks like this:

```
 <element id="Observation.status.value">
      <path value="Observation.status.value" />
      <representation value="xmlAttr" />
      <short value="Primitive value for code" />
      <definition value="Primitive value for code" />
      <min value="0" />
      <max value="1" />
      <base>
        <path value="string.value" />
        <min value="0" />
        <max value="1" />
      </base>
      <type>
        <extension url="http://hl7.org/fhir/StructureDefinition/structuredefinition-regex">
          <valueString value="[^\s]+([\s]?[^\s]+)*" />
        </extension>
        <code>
          <extension url="http://hl7.org/fhir/StructureDefinition/structuredefinition-json-type">
            <valueString value="string" />
          </extension>
          <extension url="http://hl7.org/fhir/StructureDefinition/structuredefinition-xml-type">
            <valueString value="xsd:token" />
          </extension>
          <extension url="http://hl7.org/fhir/StructureDefinition/structuredefinition-rdf-type">
            <valueString value="xsd:token" />
          </extension>
        </code>
      </type>
      <maxLength value="1048576" />
    </element> 
```

We are looking at a special case, the `Observation.status.value`.  The type of `Observation.status` is `code`, and since this element modified in the diff because of the inclusion of an extension, the whole definition of `code` is included here, including the magical `value` element of the `code` datatype. This `value` element is the magical "real primitive", it's the `value` attribute in xml. These attributes are not of a fhir type, but of an underlying low-level system type.  In this case, the `<type>` element in `ElementDefinition` (which would normally hold the FHIR type name) is now empty - it just has extensions expressing this primitive type in xml, json and rdf.

Now, if we look at the FhirPath expression, it wants the combination of `code&profile&targetProfile` to be unique. Is it unique in this case?  It's debatable since there is neither a code, nor a profile, nor a target profile present, so the concatenation of the three is empty.  I guess since there's only once "empty" value, it is still unique, but I am pretty sure the current implementation of `isDistinct()` does not support this special case.

I will start looking there. To be continued.

**Comment by ewoutkramer** [comment]
This suggests that validating the structuredefinition for the core `code` itself would also trigger this error:  

https://simplifier.net/coredatatypesstu3/code/$validate

Indeed, it does.

**Comment by ewoutkramer** [comment]
Tuned out it had nothing to do with isDistinct(), nor the & operator, but the casting logic within the FhirPath operator. As such, this issue showed a much bigger weakness in the FhirPath evaluator, than this humble "r3-only" moniker may suggest.

First. this error does not appear in R4 because in R4, the primitive `.value` property has a `<code>` again (looking something like "http://hl7.org/fhirpath/System.String"), so this evaluator bug is never triggered.

The real error only surfaced for operators and functions that do not propagate null (like `&`), in this case the operator WANTS an empty/null value to be passed into the function.

However, before we do that, we try to cast the value to the desired type. In this case, from null to string. However, the concept of a "null string" was not considerd, so the operation gave up with a cast exception.
